components/cd-dropdown/cd-dropdown.js

```diff
@@ -0,0 +1,396 @@
+    element.setAttribute('data-cd-toggable', '');
+  }
+
+  // Hide the element.
+  element.setAttribute('data-cd-hidden', expand === false);
+
+  // Remove the button fallback element if any.
+  if (element.hasAttribute('data-cd-replace')) {
+    var fallback = document.getElementById(element.getAttribute('data-cd-replace'));
+    if (fallback) {
+      fallback.parentNode.removeChild(fallback);
+    }
+    element.removeAttribute('data-cd-replace');
+  }
+
+  // Add the toggler before the toggable element if not already.
+  if (element.previousElementSibling !== toggler) {
+    element.parentNode.insertBefore(toggler, element);
+  }
+}
+
+/**
+ * Remove the element's toggler.
+ */
+function unsetToggable(element) {
+  var toggler = element.previousElementSibling;
+  if (toggler && toggler.hasAttribute('data-cd-toggler')) {
+    // Remove event handler to avoid leaking.
+    toggler.addEventListener('click', handleToggle);
+    toggler.addEventListener('keydown', handleEscape);
+    element.addEventListener('keydown', handleEscape);
+
+    // Delete toggling button.
+    toggler.parentNode.removeChild(toggler);
+
+    // Reset attributes on the toggable element.
+    element.removeEventListener('keydown', handleEscape);
+    element.removeAttribute('data-cd-hidden');
+  }
+}
+
+/**
+ * Update a toggable element, setting or removing the toggling button and
+ * attributes depending on the `dropdown` css property. When set to false
+ * we remove the toggler and reset the toggable attributes so that the HTML
+ * markup reflects the current behavior of the element.
+ */
+function updateToggable(element) {
+  if (window.getComputedStyle(element, null).getPropertyValue('--dropdown') === 'false') {
+    unsetToggable(element);
+  }
+  else {
+    setToggable(element);
+  }
+  // Mark the element as processed. This is notably used to remove the
+  // initial hidden state that is used to prevent flash of content.
+  if (!element.hasAttribute('data-cd-processed')) {
+    element.setAttribute('data-cd-processed', true);
+  }
+}
+
+/**
+ * Initialize the toggable menus, adding a toggle button and event
+ * handling.
+ */
+function initializeToggables() {
+  // Collapse dropdowns when clicking outside of the toggable target.
+  document.addEventListener('click', handleClickAway);
+
+  // Loop through the toggable elements and set/unset the toggling button
+  // depending on the screen size.
+  window.addEventListener('resize', handleResize);
+
+  // Initial setup.
+  handleResize();
+}
+
+/**
+ * Main logic.
+ */
+if (document.documentElement.classList.contains('js')) {
+
+  // Collapse popups when clicking outside of the toggable target.
+  document.addEventListener('click', handleClickAway);
+
+  document.addEventListener('click', handleEscape);
+  window.addEventListener('click', handleResize);
+  document.addEventListener('click', handleToggle);
+
+  // Initialize toggable dropdown.
+  initializeToggables();
+
+}
```

Comment:
I feel like we're polluting global scope without much need to. Many functions have unique names, but a few don't, such as `toggle`.

The component JS can at least be contained to an IIFE with its own root object, and then define each method individually on the existing object. All the event listeners just need a prefix added to work the same way. The methods can call each other using that same prefix:

```js
(function iife() {
  // Our global
  var cdDropdown = {};

  // each method gets added to the global
  cdDropdown.initializeToggables = function initializeToggables() {
    console.log('initializeToggables');
  };
  cdDropdown.calledInternally = function calledInternally() {
    // just to demo who called
    console.log('calledInternally from ' + calledInternally.caller);
  };
  cdDropdown.setToggable = function setToggable() {
    console.log('setToggable');
    // call one of the peers
    cdDropdown.calledInternally();
  };

  // set an event listener
  document.addEventListener('click', cdDropdown.setToggable);

  // init
  cdDropdown.initializeToggables();
})();
```

Do other parts of the CD try to use these methods? If they do, we can ditch the `iife()` part and contain the whole thing to `window.cdDropdown`. It doesn't look like the code stores any state, but that can also be done directly inside `cdDropdown` with properties.

Reply:
@rupl There's a note in the README https://github.com/UN-OCHA/common_design/tree/cd-226-design-system/components/cd-dropdown#caveats where I try to explain why the JS file is here.
This one is not really used, but since I was testing things outside of Drupal, and might have to support these components outside of Drupal, I've left it in place and tried to keep it up-to-date with the work done on the one we use in Drupal https://github.com/UN-OCHA/common_design/blob/cd-226-design-system/js/cd-dropdown.js
I'll do what I can to improve it

Reply:
whoops! sorry I missed the comment in the readme.

That's an interesting one. No fun to maintain two copies. Perhaps we should zero this file out and leave a comment saying "check the theme directory, or for non-drupal implementations remove the Drupal Behaviors wrapper and replace with a generic object"